### ISSUE TITLE
Cache: delta-encode cell timestamps

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#2954]
A cell's timestamp can be delta-encoded against the row marker timestamp, if it exists, or against the first cell's timestamp. This means storing the timestamps for cells at the row level. The encoding  could follow this scheme:

* If the delta is 0, just store one bit, `0`;
* If the delta is between [−8,388,608, 8,388,607], store `10` and 24 bits;
* If the delta is between [-17179869184, 17179869183], store `110` and 35 bits;
* If the delta is between [-274877906944, 274877906945], store `1110` and 39 bits;
* Else, store `11110` and 64 bits (should only be needed for users doing weird stuff);
* If the cell is a collection, store `111110`, and have the structure inside the collection cell, or in a different, support data structure.

For append-only workloads, we would be able to keep effectively a timestamp per row.

**Comment by mykaul** [comment]
@denesb , @michoecho - is that worth considering under the 'disk:mem ratio' improvements?

**Comment by nyh** [comment]
This is a seven year old issue. Is it still true? https://opensource.docs.scylladb.com/stable/architecture/sstable/sstable3/sstables-3-data-file-format.html suggests that our new sstable format ("3.0", or "mc"), added to Scylla around a year after this issue, added delta encoding of timestamps and ttls. I have only vague recollection of this issue, so maybe some with a fresher memory can remind us the current situation - or look at the code to see what it says?

**Comment by mykaul** [comment]
I thought so too - but this is talking about the cache, not sstables?

**Comment by nyh** [comment]
@mykaul oops, you're right, the sstable encoding won't be relevant to the in-memory mutation storage. I think we probably have bigger overheads in our in-memory mutation storage, but I guess the issue of timestamps can be discussed too.

**Comment by denesb** [comment]
Yes, the amount of bits taken by timestamps are the least of our concerns when it comes to in-memory storage space. We used fixed C++ data structures, which must have fixed-size integers in all cells, so clever encodings will not help us get any gains there. We'd need to come up with an in-memory serialization format first, to stand to gain from this.
